### ISSUE TITLE
学习 YOLOv5

### ISSUE THREAD

**Issue by nonocast** [nonocast/me#221]
## 体验

[ultralytics/yolov5: YOLOv5 🚀 in PyTorch > ONNX > CoreML > TFLite](https://github.com/ultralytics/yolov5)

访问官网看到Quick Start Examples部分

```shell
~ git clone https://github.com/ultralytics/yolov5  # clone
~ cd yolov5
~ pip install -r requirements.txt  # install
```

注：这里建议参考 #219 配置一下anaconda 环境，Python>=3.7.0, PyTorch>=1.7

安装后进入yolov5的目录

```shell
(taidi) ➜  yolov5 git:(master) ls data/images
bus.jpg    zidane.jpg
(taidi) ➜  yolov5 git:(master) 
(taidi) ➜  yolov5 git:(master) python3 detect.py --weights ~/Desktop/yolov5s.pt --source ./data/images/zidane.jpg 
detect: weights=['/Users/nonocast/Desktop/yolov5s.pt'], source=./data/images/zidane.jpg, data=data/coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False
YOLOv5 🚀 v6.0-235-ge1a6a0b torch 1.10.2 CPU

Fusing layers... 
[W NNPACK.cpp:79] Could not initialize NNPACK! Reason: Unsupported hardware.
Model Summary: 213 layers, 7225885 parameters, 0 gradients
image 1/1 /Users/nonocast/Develop/vendors/yolov5/data/images/zidane.jpg: 384x640 2 persons, 1 tie, Done. (0.341s)
Speed: 1.2ms pre-process, 340.7ms inference, 7.9ms NMS per image at shape (1, 3, 640, 640)
Results saved to runs/detect/exp
(taidi) ➜  yolov5 git:(master) 
```

<img src="https://user-images.githubusercontent.com/1457904/152676249-e1d0c5d0-3a23-42b4-b4de-cfb939524f4f.png">

这就完成了YOLO的第一次体验，先不要去管什么模型算法，要是看不到效果，什么都不是。

再看下视频 (--source 0):

<img src="https://user-images.githubusercontent.com/1457904/153255074-37e80aad-43e6-48b3-ae31-d74b92d67ed4.png">

然后开一个python文件，只需要5行代码

```py
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s') 

# Images
img = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, PIL, OpenCV, numpy, list

# Inference
results = model(img)

# Results
results.show()
```

上面直接通过PyTorch Hub，简单理解就是在云端提供了模型和参数的仓库供开发者直接获取。

再看下不通过PyTorch Hub的写法,

```py

```

## 理解

官网的介绍如下:

### What is YOLOv5

YOLO an acronym for 'You only look once', is an object detection algorithm that divides images into a grid system. Each cell in the grid is responsible for detecting objects within itself.

YOLO is one of the most famous object detection algorithms due to its speed and accuracy.

### The History of YOLO

**YOLOv5**

Shortly after the release of YOLOv4 Glenn Jocher introduced YOLOv5 using the Pytorch framework.
The open source code is available on [GitHub](https://github.com/ultralytics/yolov5)

Author: [Glenn Jocher](https://www.linkedin.com/in/glenn-jocher)
Released: 18 May 2020

**YOLOv4**

With the original authors work on YOLO coming to a standstill, YOLOv4 was released by Alexey Bochoknovskiy, Chien-Yao Wang, and Hong-Yuan Mark Liao. The paper was titled [YOLOv4: Optimal Speed and Accuracy of Object Detection](https://arxiv.org/abs/2004.10934v1)

Author: [Alexey Bochoknovskiy](https://ru.linkedin.com/in/alexey-bochkovskiy-1213b542), [Chien-Yao Wang](https://www.researchgate.net/profile/Chien-Yao-Wang), and [Hong-Yuan Mark Liao](https://en.wikipedia.org/wiki/Mark_Liao)
Released: 23 April 2020

**YOLOv3**

YOLOv3 improved on the YOLOv2 paper and both Joseph Redmon and Ali Farhadi, the original authors, contributed.
Together they published [YOLOv3: An Incremental Improvement](https://arxiv.org/abs/1804.02767v1)

The original YOLO papers were are hosted [here](https://pjreddie.com/darknet/yolo/)

Author: [Joseph Redmon](https://pjreddie.com/) and [Ali Farhadi](https://www.cs.washington.edu/people/faculty/ali)
Released: 8 Apr 2018

YOLOv2

YOLOv2 was a joint endevor by Joseph Redmon the original author of YOLO and Ali Farhadi.
Together they published [YOLO9000:Better, Faster, Stronger](https://arxiv.org/abs/1612.08242v1)

Author: [Joseph Redmon](https://pjreddie.com/) and [Ali Farhadi](https://www.cs.washington.edu/people/faculty/ali)
Released: 25 Dec 2016

**YOLOv1**

YOLOv1 was released as a research paper by Joseph Redmon.
The paper was titled [You Only Look Once: Unified, Real-Time Object Detection](https://arxiv.org/abs/1506.02640)

Author: [Joseph Redmon](https://pjreddie.com/)
Released: 8 Jun 2015

YOLOv5是一个family, 包括了YOLOv5n, YOLOv5s, YOLOv5m, YOLOv5l, YOLOv5x, 就是small, medium, large, xlarge的意思，YOLOv5的价值就体现在下面这张表:

<img src="https://user-images.githubusercontent.com/1457904/152676711-2deede08-d1b4-4acd-9d6b-a46d34137a41.png">

- size: 输入图像尺寸
- mAP: mean Average Precision, 目标检测的主要评价指标， 参考[https://www.zhihu.com/question/53405779](https://www.zhihu.com/question/53405779), 这里是基于COCO数据的评价指标
- Speed CPU b1: 基于CPU计算的耗时，v5s一张图片需要98ms
- Speed V100 b1: 基于Nvidia Tesla v100 ($10,000左右)，v5s一张图片只需要6.4, 是CPU版本的15分之一, 每秒可以处理150张，理论上可以做到150fps，可以做到实时处理视频，这个价值就非常大了
- Speed V100 b32: 这里的b没太明白，batch?
- Params: 模型参数尺寸，即pt文件大小
- FLOPs: Floating-point Operations Per Second， 每秒计算的浮点次数

## 扩展

// 待写

## 参考阅读

- [YOLOv5算法详解_WZZ18191171661的博客-CSDN博客_yolov5算法详解](https://blog.csdn.net/WZZ18191171661/article/details/113789486)
- [深度学习_目标检测_“YOLOv5”详解（持续更新）_Rocky6688的博客-CSDN博客_yolov5是什么东西](https://blog.csdn.net/Rocky6688/article/details/107199675)
